service/history/chasm_engine.go

```diff
@@ -0,0 +1,204 @@
+	"go.temporal.io/server/common/headers"
+	"go.temporal.io/server/common/locks"
+	"go.temporal.io/server/service/history/api"
+	"go.temporal.io/server/service/history/configs"
+	"go.temporal.io/server/service/history/consts"
+	historyi "go.temporal.io/server/service/history/interfaces"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/workflow/cache"
+)
+
+type ChasmEngine struct {
+	entityCache     cache.Cache
+	shardController shard.Controller
+	registry        *chasm.Registry
+	configs         *configs.Config
+}
+
+func NewChasmEngine(
+	entityCache cache.Cache,
+	shardController shard.Controller,
+	registry *chasm.Registry,
+	configs *configs.Config,
+) *ChasmEngine {
+	return &ChasmEngine{
+		entityCache:     entityCache,
+		shardController: shardController,
+		registry:        registry,
+		configs:         configs,
+	}
+}
+
+func (e *ChasmEngine) UpdateComponent(
+	ctx context.Context,
+	ref chasm.ComponentRef,
+	updateFn func(chasm.MutableContext, chasm.Component) error,
+	opts ...chasm.TransitionOption,
+) (updatedRef chasm.ComponentRef, retError error) {
+
+	shardContext, executionLease, err := e.getExecutionLease(ctx, ref)
+	if err != nil {
+		return chasm.ComponentRef{}, err
+	}
+	defer func() {
+		executionLease.GetReleaseFn()(retError)
+	}()
+
+	mutableState := executionLease.GetMutableState()
+	chasmTree, ok := mutableState.ChasmTree().(*chasm.Node)
+	if !ok {
+		return chasm.ComponentRef{}, serviceerror.NewInternal(
+			fmt.Sprintf("CHASM tree implementation not properly wired up, encountered type: %T", mutableState.ChasmTree()),
+		)
+	}
+
+	mutableContext := chasm.NewMutableContext(ctx, chasmTree)
+	component, err := chasmTree.Component(mutableContext, ref)
+	if err != nil {
+		return chasm.ComponentRef{}, err
+	}
+
+	if err := updateFn(mutableContext, component); err != nil {
+		return chasm.ComponentRef{}, err
+	}
+
+	// TODO: Support WithSpeculative() TransitionOption.
+
+	if err := executionLease.GetContext().UpdateWorkflowExecutionAsActive(
+		ctx,
+		shardContext,
+	); err != nil {
+		return chasm.ComponentRef{}, err
+	}
+
+	newRef, ok := mutableContext.Ref(component)
+	if !ok {
+		return chasm.ComponentRef{}, serviceerror.NewInternal(
+			fmt.Sprintf("component not found in the new compoenent tree after mutation, componentRef: %+v", ref),
+		)
+	}
+
+	return newRef, nil
+}
+
+func (e *ChasmEngine) ReadComponent(
+	ctx context.Context,
+	ref chasm.ComponentRef,
+	readFn func(chasm.Context, chasm.Component) error,
+	opts ...chasm.TransitionOption,
+) (retError error) {
+	_, executionLease, err := e.getExecutionLease(ctx, ref)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		// Always release the lease with nil error since this is a read only operation
+		// So even if it fails, we don't need to clear and reload mutable state.
+		executionLease.GetReleaseFn()(nil)
+	}()
+
+	chasmTree, ok := executionLease.GetMutableState().ChasmTree().(*chasm.Node)
```

Comment:
I don't understand this cast. Why not to change `NewContext` to accept `ChasmTree` interface (instead of `*Node`) and add missing methods to `ChasmTree`?

Reply:
> change NewContext to accept ChasmTree interface

That's a cycle dependency. `ChasmTree` is in `historyi` package which depends on `chasm`. 
That interface and noop tree impl will eventually be removed, so I guess I will just go with the simplest type assertion approach now.